ktor-client/ktor-client-core/common/src/io/ktor/client/engine/HttpClientEngine.kt

```diff
@@ -72,15 +73,26 @@ interface HttpClientEngine : CoroutineScope, Closeable {
     /**
      * Create call context and use it as a coroutine context to [execute] request.
      */
-    private suspend fun executeWithinCallContext(requestData: HttpRequestData): HttpResponseData {
+    private suspend fun executeWithinCallContext(
+        attributes: Attributes,
+        requestData: HttpRequestData
+    ): HttpResponseData {
         val callContext = createCallContext(requestData.executionContext)
 
         return async(callContext + KtorCallContextElement(callContext)) {
-            if (closed) {
+            if (this@HttpClientEngine.closed) {
                 throw ClientEngineClosedException()
             }
 
-            execute(requestData)
+            val wrapper = attributes.getOrNull(HttpClientEngineWrapper.key)
```

Comment:
Why we can't wrap the engine at the construction step?

Reply:
Well, I changed the approach and now we wrap it at the construction step.